test/powershell/Language/Parser/MethodInvocation.Tests.ps1

```diff
@@ -81,3 +81,196 @@ namespace MSFT_716893
+                ErrorCount     = 2
+            }
+        )
+    }
+
+    It 'does not throw a parse error for "<Script>"' -TestCases $EmptyArrayCases {
+        param($Script)
+
+        { [scriptblock]::Create($script) } | Should -Not -Throw
+    }
+
+    It 'reports a parse error for "<Script>"' -TestCases $ExpectedParseErrors {
+        param($Script, $ExpectedError, $ErrorCount)
+
+        $parseErrors = $null
+        [System.Management.Automation.Language.Parser]::ParseInput($Script, [ref]$null, [ref]$parseErrors)
+
+        $parseErrors.Count | Should -Be $ErrorCount
+        $parseErrors.ErrorId | Should -BeIn $ExpectedError
+    }
+
+    It 'can call a generic method "<Script>" with no arguments' -TestCases $EmptyArrayCases {
+        param($Script, $ExpectedType)
+
+        $Result = & [scriptblock]::Create($Script)
+        $Result.GetType() | Should -Be $ExpectedType
+
+        $Result.Length | Should -Be 0
+    }
+
+    It 'can call generic instance methods' {
+        $dictionary = [System.Collections.Concurrent.ConcurrentDictionary[string, int]]::new()
+
+        $addEntryScript = {
+            param($key, $float)
+
+            if ($float -gt 0.5) {
+                return 10
+            }
+            else {
+                return 1
+            }
+        }
+
+        $updateEntryScript = {
+            param($key, $currentValue, $float)
+
+            if ($currentValue / $float -gt 2) {
+                return 5
+            }
+            else {
+                return 0
+            }
+        }
+
+        $FloatValue = 0.4
+        $Key = 'Test'
+
+        # Add entry
+        $dictionary.AddOrUpdate[float]($Key, $addEntryScript, $updateEntryScript, $FloatValue)
+        $dictionary.$Key | Should -Be 1
+
+        # Update entry
+        $dictionary.AddOrUpdate[float]($Key, $addEntryScript, $updateEntryScript, $FloatValue)
+        $dictionary.$Key | Should -Be 5
+    }
+
+    It 'can call generic static methods with arguments' {
+        [System.Linq.Enumerable]::Select[int, int](
+            [int[]](0..10),
+            [func[int, int]] { $args[0] + 2 }
+        ) | Should -Be @(2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12)
+    }
+
+    It 'gives a runtime error if too many type parameters are given' {
+        try {
+            [array]::Empty[string, int]()
+        }
+        catch {
+            $_.Exception.Message | Should -BeExactly 'Cannot find an overload for "Empty" and the argument count: "0".'
+            $_.FullyQualifiedErrorId | Should -BeExactly 'MethodCountCouldNotFindBest'
+        }
+    }
+
+    It 'gives a runtime error if a nonexistent type is specified' {
+        try {
+            [array]::Empty[thisdoesnotexist]()
+        }
+        catch {
+            $_.Exception.Message | Should -BeExactly 'Cannot find an overload for "empty" and the argument count: "0".'
+            $_.FullyQualifiedErrorId | Should -BeExactly 'MethodCountCouldNotFindBest'
+        }
+    }
```

Comment:
@rjmholt these two are the ones I like the least. I'd like there to be more actionable errors here, such as:

- `type specified for generic method was not found`
- `a valid generic method overload that takes the specified number of type arguments was not found`

Thus far though, I've not had much luck with incorporating logic for this kind of case. 😕

Reply:
I managed to find some... okay... solutions here. Part of the problem getting a reasonable error message for something like `[array]::Empty[notexistent]()` is that a nonexistent type name doesn't leave retain its original lexeme at this point in the invocation, we just have a `null` entry to work with.

This is still miles better than the default error message that shows up here, though.